Fullview Windows

linusha opened the pull request:
There are multiple things in this:

1. I added the possibility to react to DoubleClicks to `Morphs`. This does not use the `dblclick` event of the dom, but is in line with how one currently reacts to clicks.

2. I fixed resizing of windows via commands, before they only updated their inner layout after another interaction with the window, which led to weird looking windows directly after resizing.

3. A double-click on the title of a window now maximizes the window in a way that respects the actually visible part of the world, i.e. does not overlap with the top-bar.

**I have a few open questions that came up while working on this:**

1. Currently a window does not have a resizer at the top. Is there a specific reason for this or should we add one?
2. After playing around with the options for resizing windows some more I realized, that I think all options should always respect the topbar. Also, all options should respect the currently visible part of the world, i.e. in the case relevant for the new feature, if the world is two screens high and I double-click a window at the bottom of the world, it should fill the "lower screen" and not jump to the top. The new option I introduced would be obsolete and equivalent to "full" and "fullscreen".

Please let me know what you think about this @merryman, then I will correct the change in this PR and double-check all exsiting options.

merryman commented:
> 1. I added the possibility to react to DoubleClicks to `Morphs`. This does not use the `dblclick` event of the dom, but is in line with how one currently reacts to clicks.

The DOM event system is utterly broken, which is why we try to stay away from it as much as we can. In that sense this just the right thing to do.

> 2. I fixed resizing of windows via commands, before they only updated their inner layout after another interaction with the window, which led to weird looking windows directly after resizing.

Good call, maybe it's worthwhile overridding `setBounds()` for the window and trigger the relayout routines in there by default.

> 3. A double-click on the title of a window now maximizes the window in a way that respects the actually visible part of the world, i.e. does not overlap with the top-bar.

Looking good!
 
> 1. Currently a window does not have a resizer at the top. Is there a specific reason for this or should we add one?

Nope, initially there was just a resizer located on the bottom left which I expanded over time. For some reason I left out the top, but feel free to add that in here as well.

> 2. After playing around with the options for resizing windows some more I realized, that I think all options should always respect the topbar. Also, all options should respect the currently visible part of the world, i.e. in the case relevant for the new feature, if the world is two screens high and I double-click a window at the bottom of the world, it should fill the "lower screen" and not jump to the top. The new option I introduced would be obsolete and equivalent to "full" and "fullscreen".

Agreed. I forgot adjusting these commands after I introduced the top bar.

linusha commented:
> Good call, maybe it's worthwhile overridding `setBounds()` for the window and trigger the relayout routines in there by default.

Good call, I did that.

> > 1. A double-click on the title of a window now maximizes the window in a way that respects the actually visible part of the world, i.e. does not overlap with the top-bar.
> 
> Looking good!

I also use the existing, but seeminlgy unused `maximized` property now. This way the first double-click maximizes the window, while the second uses the `reset` option to make it smaller again (and so forth).
 
> > 1. Currently a window does not have a resizer at the top. Is there a specific reason for this or should we add one?
> 
> Nope, initially there was just a resizer located on the bottom left which I expanded over time. For some reason I left out the top, but feel free to add that in here as well.

I added the three remaining ones.

> > 1. After playing around with the options for resizing windows some more I realized, that I think all options should always respect the topbar. Also, all options should respect the currently visible part of the world, i.e. in the case relevant for the new feature, if the world is two screens high and I double-click a window at the bottom of the world, it should fill the "lower screen" and not jump to the top. The new option I introduced would be obsolete and equivalent to "full" and "fullscreen".
> 
> Agreed. I forgot adjusting these commands after I introduced the top bar.

I was very brutal when editing that code, i.e. I basically eliminated a lot of options that I could not see anyone using. Feel absolutely free to tell me if you think there is value in keeping them (I could leave them as comments or fix them properly, as you wish). If we decide to eliminate them I would also clean up the keybindings before merging this.

linusha commented on lively.components/window.js:
```diff
@@ -88,7 +88,12 @@ export default class Window extends Morph {
+      maximized: {
+        set(isMaximized) {
+          this.setProperty('maximized', isMaximized);
+          this.toggleMaximize;
```
Good catch! I actually opted to write this in a style similar to how `minimized` is handled, as these two should be consistent.

I did `aWindow.copy()` both for a maximized and non-maximized window and both times the result looked good in an inspector.

linusha commented on lively.morphic/world.js:
```diff
@@ -302,12 +302,20 @@ export class World extends Morph {
 
+  fullVisibleBounds() {
+    // returns the visible rect of the world with respect to the topbar
+    let bar = $world.getSubmorphNamed("lively top bar");
```
That is good to know! I moved the method and added a check for existence of the `top bar` Morph. While I was at it I also used more descriptive names.

merryman commented:
Please resolve the conflicts here, so we can proceed.

linusha commented:
> Please resolve the conflicts here, so we can proceed.

Please consider #109 instead. Sorry for the inconvenience and the delay! :see_no_evil: Thank you!
